<a href="https://colab.research.google.com/github/rushil00/ML-and-AI-Practice/blob/main/Module5jup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 5

Graphical user interfaces, event-driven programming paradigm, tkinter module, creating simple GUI, buttons, labels, entry fields, dialogs, widget attributes - sizes, fonts, colors layouts, nested frames, Multithreading, Networks, and Client/Server Programming, introduction to HTML, interacting with remote 
HTML server, running html-based queries, downloading pages; 
CGI programming, programming a simple CGI form

# Graphical User Interface (GUI)


Python has no native support for GUI programming, but several GUI libraries
can be used by Python programs. The Tk library is available using the tkinter
module, and is usually installed as standard. 


GUI programming is a bit different from writing console programs.


*   Python console programs and module files always have a .py extension, but
for Python GUI programs we use a .pyw extension (module files always use .py,
though).


*    On Windows, .pyw ensures that Windows uses the pythonw.exe interpreter instead of python.exe, and this in turn ensures that when we execute a Python GUI program, no unnecessary console window will appear.

# Brief inventory of GUI toolkits available to Python programmers 
<dl>
<dt>tkinter 
<dd>An open source GUI library and the continuing de facto standard for portable GUI development in Python. Python scripts that use tkinter to build GUIs run portably on Windows, X Windows (Unix and Linux), and Macintosh OS X, and they display a native look-and-feel on each of these platforms today. 
<dt>wxPython
<dd>A Python interface for the open source wxWidgets (formerly called wxWindows)
library, which is a portable GUI class framework originally written to be used from the C++ programming language. 
<dt>PyQt
<dd>A Python interface to the Qt toolkit (now from Nokia, formerly by Trolltech), and perhaps the third most widely used GUI toolkit for Python today. PyQt is a full featured GUI library and runs portably today on Windows, Mac OS X, and Unix and Linux.
<dt>PyGTK
<dd>A Python interface to GTK, a portable GUI library originally used as the core of the Gnome window system on Linux. The gnome-python and PyGTK extension
packages export Gnome and GTK toolkit calls.

In [ ]:
import tkinter

In [ ]:
tkinter.TkVersion

8.6

In [ ]:
from tkinter import Label # Loads a widget class from the tkinter module

In [ ]:
from tkinter import Tk

In [ ]:
root = Tk()
Label(None, text='Hello GUI world!',bg='red',fg='white').pack(expand = Yes, fill ='both')
root.mainloop()

In [ ]:
widget = Label(None, text='Hello GUI world!') #  Makes an instance of the imported Label class
widget.place(relx=.2, rely=.2) #  Packs (arranges) the new Label in its parent widget
widget.mainloop() # Calls mainloop to bring up the window and start the tkinter event loop

 In fact, the tkinter mainloop function is similar in spirit to the following pseudo-Python code:
```
def mainloop():
    while the main window has not been closed:
         if an event has occurred:
             run the associated event handler function
```

Because of this model, the mainloop call in above Example never returns to our script while the GUI is displayed on-screen.‡ 
<br>When we write larger scripts, the only way we can get anything done after calling mainloop is to register callback handlers to respond to events.
<br>This is called event-driven programming, and it is perhaps one of the most unusual aspects of GUIs. GUI programs take the form of a set of event handlers that share saved information rather than of a single main control flow. 

In [ ]:
from tkinter import *

In [ ]:
root = Tk()
Label(root, text='Hello GUI world!').pack(side=LEFT)
root.mainloop()

# Widget Resizing Basics

Top-level windows, such as the one built by all of the coding variants we have seen thus far, can normally be resized by the user; simply drag out the window with your mouse. 
<br>This isn’t very good—the label stays attached to the top of the parent window instead of staying in the middle on expansion—but it’s easy to improve on this with a pair of pack options, demonstrated as below.


In [ ]:
from tkinter import *
Label(text='Hello GUI world!').pack(expand=YES, fill=BOTH)
mainloop()

The text option of the label is set after it is constructed, by assigning to the widget’s text key. 
<br>Widget objects overload (intercept) index operations such that options are also available as mapping keys, much like a dictionary.

In [ ]:
from tkinter import *
widget = Label()
widget['text'] = 'Hello GUI world!'
widget.pack(side=TOP)
mainloop()

More commonly, widget options can be set after construction by calling the widget config method

In [ ]:
from tkinter import *
root = Tk()
widget = Label(root)
widget.config(text='Hello GUI world!')
widget.pack(side=TOP, expand=YES, fill=BOTH)
root.title('gui1g.py')
root.mainloop()


# Adding Buttons and Callbacks

Here, instead of making a label, we create an instance of the tkinter Button class. It’s attached to the default top level window as before on the default TOP packing side. But the main thing to notice here is the button’s configuration arguments: we set an option called command to the ```sys.exit``` function

In [ ]:
import sys
from tkinter import *
root = Tk()
widget = Button(root, text='Close Window', command=root.destroy) # command=sys.exit
widget.pack(side=BOTTOM)
root.mainloop()

Annother version that packs the button in place without assigning it to a name, attaches it to the LEFT side of its parent window explicitly, and specifies ```root.quit``` as the callback handler—a standard Tk object method that shuts down the GUI and so ends the program.

In [ ]:
from tkinter import *
root = Tk()
Button(root, text='press', command=root.quit).pack(side=LEFT)
root.mainloop()


If you want the button to be given all available space and to stretch to fill all of its
assigned space horizontally, add expand=YES and fill=X keyword arguments to the
pack call.

In [ ]:
from tkinter import *
root = Tk()
Button(root, text='press', command=root.quit).pack(side=LEFT,expand=YES, fill=X)
root.mainloop()

In [ ]:
from tkinter import *
root = Tk()
Button(root, text='press', command=root.quit).pack(side=LEFT,expand=YES, fill=Y)
root.mainloop()

In [ ]:
from tkinter import *
root = Tk()
Button(root, text='press', command=root.quit).pack(side=LEFT,expand=YES, fill=BOTH)
root.mainloop()

# Adding User-Defined Callback Handlers
The command option specifies a function
we’ve defined locally. When the button is pressed, tkinter calls the quit function in this
file to handle the event, passing it zero arguments.

In [ ]:
import sys
from tkinter import *
def quit(): # a custom callback handler
    print('Hello, I must be going...') # kill windows and process
    root.destroy()
 

root =Tk()
widget = Button(None, text='Hello event world', command=quit)
widget.pack()
widget.mainloop()

Hello, I must be going...


In [ ]:
import tkinter as tk
root = tk.Tk()
root.geometry("100x50")

def close_window():
    root.destroy()

button = tk.Button(text = "Click and Quit", command = close_window)
button.pack()
button.place(x = 20, y = 30)

root.mainloop()

In [ ]:
#Import tkinter library
from tkinter import *
#Create an instance of Tkinter frame or window
win= Tk()
#Set the geometry of tkinter frame
win.geometry("750x250")
def callback():
    Label(win, text="Hello World!", font=('Century 20 bold')).pack(pady=0)
#Create a Label and a Button widget
btn=Button(win, text="Press Enter", command= callback)
btn.pack(ipadx=10)
win.bind('<Return>',lambda event:callback())
win.mainloop()

# Configuring Widget Appearance
 tkinter widgets can be made to look arbitrarily different, though, using a handful of widget and packer options

In [ ]:
from tkinter import *
root = Tk()
labelfont = ('times', 20, 'bold') # family, size, style
widget = Label(root, text='Hello config world')
widget.config(bg='black', fg='yellow') # yellow text on black label
widget.config(font=labelfont) # use a larger font
widget.config(height=3, width=30) # initial size: lines,chars
widget.pack(expand=YES)
root.mainloop()

In [ ]:
from tkinter import *
widget = Button(text='Spam', padx=10, pady=10) #Extra space can be added around many widgets with the padx= N and pady= N options.
widget.pack(padx=20, pady=20)
widget.config(cursor='gumby') # A cursor = { gumby,watch, pencil, cross, hand2} option can be given to change the 
                #appearance of the mouse pointer when it moves over the widget.
widget.config(bd=8, relief=RAISED) #A bd= N widget option can be used to set border width, 
                 # and a relief= {FLAT, SUNKEN, RAISED, GROOVE, SOLID, RIDGE} option can specify a border style
widget.config(bg='dark green', fg='white')
widget.config(font=('helvetica', 20, 'underline italic'))
mainloop()


# Dialogs
Dialogs are windows popped up by a script to provide or request additional information. They come in two flavors, modal and nonmodal:
<dl>
    <dt>Modal
    <dd>These dialogs block the rest of the interface until the dialog window is dismissed;users must reply to the dialog before the program continues.
    <dt>Nonmodal
    <dd>These dialogs can remain on-screen indefinitely without interfering with other windows in the interface; they can usually accept inputs at any time.
</dl>


There are essentially three ways to present pop-up dialogs to users with tkinter—
<li>by using common dialog calls, 
<li>by using the now-dated Dialog object, and 
<li>by creating custom dialog windows with Toplevels and other kinds of widgets.

### Standard (Common) Dialogs
tkinter comes with a collection of precoded dialog windows that implement many of the most common pop ups programs generate—<li>file selection dialogs,<li>error and warning pop ups, and <li>question and answer prompts. 

<br>They are called standard dialogs (and sometimes common dialogs) because they are part of the tkinter library, and they use platform-specific library calls to look like they should on each platform. 

In [ ]:
from tkinter import *
from tkinter.messagebox import *
def callback():
 if askyesno('Verify', 'Do you really want to quit?'):
     showwarning('Yes', 'Quit not yet implemented')
 else:
     showinfo('No', 'Quit has been cancelled')
errmsg = 'Sorry, no Spam allowed!'
Button(text='Quit', command=callback).pack(fill=X)
Button(text='Spam', command=(lambda: showerror('Spam', errmsg))).pack(fill=X)
mainloop()


###  The Old-Style Dialog Module
In older Python code, you may see dialogs occasionally coded with the standard tkinter dialog module. This is a bit dated now, and it uses an X Windows look-and-feel; but just in case you run across such code in your Python maintenance excursions

In [ ]:
from tkinter import *
from tkinter.dialog import Dialog
class OldDialogDemo(Frame):
    def __init__(self, master=None):
        Frame.__init__(self, master)
        Pack.config(self) # same as self.pack()
        Button(self, text='Pop1', command=self.dialog1).pack()
        Button(self, text='Pop2', command=self.dialog2).pack()
    def dialog1(self):
        ans = Dialog(self, title = 'Popup Fun!', text = 'An example of a popup-dialog box, using older "Dialog.py".',
            bitmap = 'questhead', default = 0, strings = ('Yes', 'No', 'Cancel'))
        if ans.num == 0: 
            self.dialog2()
    def dialog2(self):
        Dialog(self, title = 'HAL-9000', text = "I'm afraid I can't let you do that, Dave...", bitmap = 'hourglass',
               default = 0, strings = ('spam', 'SPAM'))


if __name__ == '__main__': OldDialogDemo().mainloop()

### Custom dialogs 
Custom dialogs support arbitrary interfaces, but they are also the most complicated to program. Even so, there’s not much to it—simply create a pop-up window as a Toplevel with attached widgets, and arrange a callback handler to fetch user inputs
entered in the dialog (if any) and to destroy the window. To make such a custom dialog modal, we also need to wait for a reply by giving the window input focus, making other windows inactive, and waiting for an event

In [ ]:
import sys
from tkinter import *
makemodal = (len(sys.argv) > 1)
def dialog():
    win = Toplevel() # make a new window
    Label(win, text='Hard drive reformatted!').pack() # add a few widgets
    Button(win, text='OK', command=win.destroy).pack() # set destroy callback
    if makemodal:
        win.focus_set() # take over input focus,
        win.grab_set() # disable other windows while I'm open,
        win.wait_window() # and wait here until win destroyed
        print('dialog exit') # else returns right away
root = Tk()
Button(root, text='popup', command=dialog).pack(fill='x')
root.mainloop()

dialog exit


# Binding Events
Bind is commonly used in conjunction with other widgets (e.g., to catch return key presses for input boxes)

In [ ]:
from tkinter import *
def showPosEvent(event):
    print('Widget=%s X=%s Y=%s' % (event.widget, event.x, event.y))

def showAllEvent(event):
    print(event)
    for attr in dir(event):
        if not attr.startswith('__'):
            print(attr, '=>', getattr(event, attr))

def onKeyPress(event):
    print('Got key press:', event.char)
    
def onArrowKey(event):
    print('Got up arrow key press')
    
def onReturnKey(event):
    print('Got return key press')
    
def onLeftClick(event):
    print('Got left mouse button click:', end=' ')
    showPosEvent(event)
    
def onRightClick(event):
    print('Got right mouse button click:', end=' ')
    showPosEvent(event)
    
def onMiddleClick(event):
    print('Got middle mouse button click:', end=' ')
    showPosEvent(event)
    showAllEvent(event)
    
def onLeftDrag(event):
    print('Got left mouse button drag:', end=' ')
    showPosEvent(event)
    
def onDoubleLeftClick(event):
    print('Got double left mouse click', end=' ')
    showPosEvent(event)
    tkroot.destroy()

tkroot = Tk()
labelfont = ('courier', 20, 'bold') # family, size, style
widget = Label(tkroot, text='Hello bind world')
widget.config(bg='red', font=labelfont) # red background, large font
widget.config(height=5, width=20) # initial size: lines,chars
widget.pack(expand=YES, fill=BOTH)

widget.bind('<Button-1>', onLeftClick) # mouse button clicks
widget.bind('<Button-3>', onRightClick)
widget.bind('<Button-2>', onMiddleClick) # middle=both on some mice
widget.bind('<Double-1>', onDoubleLeftClick) # click left twice
widget.bind('<B1-Motion>', onLeftDrag) # click left and move

widget.bind('<KeyPress>', onKeyPress) # all keyboard presses
widget.bind('<Up>', onArrowKey) # arrow button pressed
widget.bind('<Return>', onReturnKey) # return/enter key pressed
widget.focus() # or bind keypress to tkroot
tkroot.title('Click Me')
tkroot.mainloop()


Got left mouse button click: Widget=.!label X=68 Y=43
Got left mouse button drag: Widget=.!label X=69 Y=43
Got left mouse button drag: Widget=.!label X=70 Y=43
Got left mouse button drag: Widget=.!label X=71 Y=43
Got left mouse button drag: Widget=.!label X=72 Y=44
Got left mouse button drag: Widget=.!label X=73 Y=45
Got left mouse button drag: Widget=.!label X=74 Y=46
Got left mouse button drag: Widget=.!label X=75 Y=47
Got left mouse button drag: Widget=.!label X=77 Y=48
Got left mouse button drag: Widget=.!label X=79 Y=49
Got left mouse button drag: Widget=.!label X=80 Y=50
Got left mouse button drag: Widget=.!label X=81 Y=51
Got left mouse button drag: Widget=.!label X=82 Y=52
Got left mouse button drag: Widget=.!label X=83 Y=53
Got left mouse button drag: Widget=.!label X=84 Y=54
Got left mouse button drag: Widget=.!label X=85 Y=54
Got left mouse button drag: Widget=.!label X=86 Y=56
Got left mouse button drag: Widget=.!label X=87 Y=57
Got left mouse button drag: Widget=.!label X=

Got left mouse button drag: Widget=.!label X=130 Y=72
Got left mouse button drag: Widget=.!label X=132 Y=71
Got left mouse button drag: Widget=.!label X=134 Y=69
Got left mouse button drag: Widget=.!label X=136 Y=68
Got left mouse button drag: Widget=.!label X=137 Y=67
Got left mouse button drag: Widget=.!label X=138 Y=67
Got left mouse button drag: Widget=.!label X=140 Y=66
Got left mouse button drag: Widget=.!label X=141 Y=65
Got left mouse button drag: Widget=.!label X=143 Y=64
Got left mouse button drag: Widget=.!label X=145 Y=62
Got left mouse button drag: Widget=.!label X=146 Y=62
Got left mouse button drag: Widget=.!label X=147 Y=60
Got left mouse button drag: Widget=.!label X=149 Y=59
Got left mouse button drag: Widget=.!label X=150 Y=58
Got left mouse button drag: Widget=.!label X=151 Y=58
Got left mouse button drag: Widget=.!label X=151 Y=57
Got left mouse button drag: Widget=.!label X=151 Y=56
Got left mouse button drag: Widget=.!label X=152 Y=56
Got left mouse button drag: 

# Message
The Message widget is simply a place to display text. Although the standard showinfo dialog we met earlier is perhaps a better way to display pop-up messages, Message splits up long strings automatically and flexibly and can be embedded inside container widgets any time you need to add some read-only text to a display. 

In [ ]:
from tkinter import *
msg = Message(text="Oh by the way, which one's Pink?")
msg.config(bg='pink', font=('times', 16, 'italic'))
msg.pack(fill=X, expand=YES)
mainloop()

# Entry
The Entry widget is a simple, single-line text input field. It is typically used for input
fields in form-like dialogs and anywhere else you need the user to type a value into a
field of a larger display. Entry also supports advanced concepts such as scrolling, key
bindings for editing, and text selections, but it’s simple to use in practice

In [ ]:
%%writefile quitter.py
"""
a Quit button that verifies exit requests;
to reuse, attach an instance to other GUIs, and re-pack as desired
"""
from tkinter import * # get widget classes
from tkinter.messagebox import askokcancel # get canned std dialog
class Quitter(Frame): # subclass our GUI
    def __init__(self, parent=None): # constructor method
        Frame.__init__(self, parent)
        self.pack()
        widget = Button(self, text='Quit', command=self.quit)
        widget.pack(side=LEFT, expand=YES, fill=BOTH)
        
    def quit(self):
        ans = askokcancel('Verify exit', "Really quit?")
        if ans: Frame.destroy(self)
            
            
if __name__ == '__main__': Quitter().mainloop()



Overwriting quitter.py


In [ ]:
from tkinter import *
from quitter import Quitter
root = Tk()
Quitter(root).pack(side=RIGHT)
root.mainloop()

In [ ]:
from tkinter import *
from quitter import Quitter
def fetch():
    print('Input => "%s"' % ent.get()) # get text
    
    
root = Tk()
ent = Entry(root)
ent.insert(0, 'Type words here') # set text
ent.pack(side=TOP, fill=X) # grow horiz
ent.focus() # save a click
ent.bind('<Return>', (lambda event: fetch())) # on enter key
btn = Button(root, text='Fetch', command=fetch) # and on button
btn.pack(side=LEFT)
Quitter(root).pack(side=RIGHT)
root.mainloop()


Input => "ashok"
Input => "ashok"


## Checkbuttons
The Checkbutton and Radiobutton widgets are designed to be associated with tkinter variables: clicking the button changes the value of the variable, and setting the variable changes the state of the button to which it is linked. In fact, tkinter variables are central to the operation of these widgets:
<li> A collection of Checkbuttons implements a multiple-choice interface by assigning each button a variable of its own.
<li> A collection of Radiobuttons imposes a mutually exclusive single-choice model by giving each button a unique value and the same tkinter variable.

In [ ]:
from tkinter import *
from tkinter import messagebox
import tkinter

top = tkinter.Tk()
CheckVar1 = IntVar()
CheckVar2 = IntVar()
C1 = Checkbutton(top, text = "Music", variable = CheckVar1, \
                 onvalue = 1, offvalue = 0, height=5, \
                 width = 20)
C2 = Checkbutton(top, text = "Video", variable = CheckVar2, \
                 onvalue = 1, offvalue = 0, height=5, \
                 width = 20)
C1.pack()
C2.pack()

top.mainloop()
print(CheckVar1.get())
print(CheckVar2.get())

1
1


In [ ]:
# check buttons, the hard way (without variables)
from tkinter import *
states = [] # change object not name
def onPress(i): # keep track of states
    states[i] = not states[i] # changes False->True, True->False

root = Tk()
for i in range(10):
    chk = Checkbutton(root, text=str(i), command=(lambda i=i: onPress(i)) )
    chk.pack(side=LEFT)
    states.append(False)
root.mainloop()
print(states) # show all states on exit

[False, False, True, False, False, False, True, False, False, True]


In [ ]:
%%writefile dialogTable.py
# define a name:callback demos table
from tkinter.filedialog import askopenfilename # get standard dialogs
from tkinter.colorchooser import askcolor # they live in Lib\tkinter
from tkinter.messagebox import askquestion, showerror
from tkinter.simpledialog import askfloat
demos = {
 'Open': askopenfilename,
 'Color': askcolor,
 'Query': lambda: askquestion('Warning', 'You typed "rm *"\nConfirm?'),
 'Error': lambda: showerror('Error!', "He's dead, Jim"),
 'Input': lambda: askfloat('Entry', 'Enter credit card number')
}

Overwriting dialogTable.py


In [ ]:
"create a bar of check buttons that run dialog demos"
from tkinter import * # get base widget set
from dialogTable import demos # get canned dialogs
from quitter import Quitter # attach a quitter object to "me"
class Demo(Frame): 
    def __init__(self, parent=None, **options):
        Frame.__init__(self, parent, **options)
        self.pack()
        self.tools()
        Label(self, text="Check demos").pack()
        self.vars = []
        for key in demos:
            var = IntVar()
            Checkbutton(self, text=key, variable=var, command=demos[key]).pack(side=LEFT)
            self.vars.append(var)
            
    def report(self):
        for var in self.vars:
            print(var.get(), end=' ') # current toggle settings: 1 or 0
            print()
    def tools(self):
        frm = Frame(self)
        frm.pack(side=RIGHT)
        Button(frm, text='State', command=self.report).pack(fill=X)
        Quitter(frm).pack(fill=X)
        
if __name__ == '__main__': Demo().mainloop()


0 
0 
1 
1 
1 


## Radio Buttons
Radio buttons are toggles too, but they are generally used in groups: just like the mechanical station selector pushbuttons on radios of times gone by, pressing one Radio button widget in a group automatically deselects the one pressed last. In other words, at most, only one can be selected at one time. 

In [ ]:
# see what happens when some buttons have same value
from tkinter import *
root = Tk()
var = StringVar()
for i in range(10):
    rad = Radiobutton(root, text=str(i), variable=var, value=str(i % 3))
    rad.pack(side=LEFT)
    var.set(' ') # deselect all initially
root.mainloop()

In [ ]:
# radio buttons, the easy way
from tkinter import *
root = Tk() # IntVars work too
var = IntVar() # select 0 to start
for i in range(10):
    rad = Radiobutton(root, text=str(i), value=i, variable=var)
    rad.pack(side=LEFT)
root.mainloop()
print(var.get()) # show state on exit

4


## Scales (Sliders)
Scales (sometimes called “sliders”) are used to select among a range of numeric values.
Moving the scale’s position with mouse drags or clicks moves the widget’s value among
a range of integers and triggers Python callbacks if registered.


In [ ]:
from tkinter import *
root = Tk()
scl = Scale(root, from_=-100, to=100, tickinterval=50, resolution=10)
scl.pack(expand=YES, fill=Y)
def report():
    print(scl.get())
Button(root, text='state', command=report).pack(side=RIGHT)
root.mainloop()


0
60
-90


In [ ]:
# Design a simple calculator using Tkinter

# Top-Level Windows
Each Toplevel object created produces a new window on the display and automatically
adds it to the program’s GUI event-loop processing stream (you don’t need to call the
mainloop method of new windows to activate them). 

In [ ]:
import sys
from tkinter import Toplevel, Button, Label
win1 = Toplevel() # two independent windows
win2 = Toplevel() # but part of same process
Button(win1, text='Spam', command=sys.exit).pack()
Button(win2, text='SPAM', command=sys.exit).pack()
Label(text='Popups').pack() # on default Tk() root window
win1.mainloop()

# Menus
Menus are the pull-down lists you’re accustomed to seeing at the top of a window (or 
the entire display, if you’re accustomed to seeing them that way on a Macintosh). Move 
the mouse cursor to the menu bar at the top and click on a name (e.g., File), and a list 
of selectable options pops up under the name you clicked (e.g., Open, Save). 

It works like this:
1. Create a topmost Menu as the child of the window widget and configure the window’s menu attribute to be the new Menu.
2. For each pull-down object, make a new Menu as the child of the topmost Menu and 
add the child as a cascade of the topmost Menu using add_cascade.
3. Add menu selections to each pull-down Menu from step 2, using the command options 
of add_command to register selection callback handlers.
4. Add a cascading submenu by making a new Menu as the child of the Menu the cascade 
extends and using add_cascade to link the parent to the child.

In [ ]:
from tkinter import * # get widget classes
from tkinter.messagebox import * # get standard dialogs
def notdone():
 showerror('Not implemented', 'Not yet available')
def makemenu(win):
 top = Menu(win) # win=top-level window
 win.config(menu=top) # set its menu option
 file = Menu(top)
 file.add_command(label='New...', command=notdone, underline=0)
 file.add_command(label='Open...', command=notdone, underline=0)
 file.add_command(label='Quit', command=win.quit, underline=0)
 top.add_cascade(label='File', menu=file, underline=0)
 edit = Menu(top, tearoff=False)
 edit.add_command(label='Cut', command=notdone, underline=0)
 edit.add_command(label='Paste', command=notdone, underline=0)
 edit.add_separator()
 top.add_cascade(label='Edit', menu=edit, underline=0)
 submenu = Menu(edit, tearoff=True)
 submenu.add_command(label='Spam', command=win.quit, underline=0)
 submenu.add_command(label='Eggs', command=notdone, underline=0)
 edit.add_cascade(label='Stuff', menu=submenu, underline=0)
if __name__ == '__main__':
    # is used to execute some code only if the file was run directly, and not imported
 root = Tk() # or Toplevel()
 root.title('menu_win') # set window-mgr info
 makemenu(root) # associate a menu bar
 msg = Label(root, text='Window menu basics') # add something below
 msg.pack(expand=YES, fill=BOTH)
 msg.config(relief=SUNKEN, width=40, height=7, bg='beige')
 root.mainloop()

# Multithreading 
Improving and making code faster is the next step after the basic knowledge of Python is acquired. Multithreading is one such way to achieve that optimization using “Threads”. 


##### https://www.upgrad.com/blog/multithreading-in-python/

### Threads in Python
When we think multitasking, we think parallel execution. Multithreading not strictly parallel execution. Threads can be thought of as separate entities of execution flow of different parts of your program running independently. So essentially, threads do not execute in parallel, but Python switches from one thread to another so fast that it seems they are parallel.

Processes on the other hand are strictly parallel and run on different cores to achieve execution faster. Threads can also be run on different processors, but they still won’t be running in parallel technically.

<b> Multithreading don’t always make processing faster. Multithreading is specifically used in tasks where threads will make processing faster.</b>

All the information of a thread is contained in the Thread Control Block(TCB). TCB consists of the following main parts:

<li>A unique TID – Thread Identifier</li>
<li>Stack Pointer which points to thread’s stack in the process </li>
<li>A Program counter which stores the address of the instruction currently being executed by the thread</li>
<li>State of the Thread (running, ready, waiting, start or done)</li>

Having said that, processes can contain multiple threads in it which share the code, data and all the files. And all the threads have their own separate register and stack which they have access to.

Now you might wonder, if the threads use the common data and code, how can they all use it without hampering other threads. This is the biggest limitation of Multithreading.

<b>Context Switching</b>

Now as described above, threads don’t run parallely, but consequently. So when one thread T1 starts execution, all other threads remain in waiting mode. Only after T1 is done with its execution can any other queued thread begin to execute. Python switches from one thread to another so fast that it seems like parallel execution. This switching is what we call ‘Context Switching’.

In [ ]:
import threading

def cuber(n):
    print("Cube: {}".format(n * n * n))

def squarer(n):
    print("Square: {}".format(n * n))

if __name__ == "__main__":
    # create the thread
    t1 = threading.Thread(target=squarer, args=(5,))
    t2 = threading.Thread(target=cuber, args=(5,))

    # start the thread t1
    t1.start()
    # start the thread t2
    t2.start()

    # wait until t1 is completed
    t1.join()
    # wait until t2 is completed
    t2.join()

    # both threads completed
    print("Done!") 

Square: 25
Cube: 125
Done!


### Thread Synchronization

As we discussed above, threads do not execute in parallel, instead Python switches from one to another. So, there is a very critical need of correct synchronization between the threads to avoid any weird behavior. 

### Race Condition

Threads which are under the same process use common data and files which can lead to a “Race” for the data between multiple threads. Therefore, if a piece of data is accessed by multiple threads, it will be modified by both the threads and the results we’ll get won’t be as expected. This is called a Race Condition.

The aim of Thread Synchronization is to make sure this Race Condition never comes and the critical section of code is accessed by threads one at a time in a synchronized way.

### Locks

To solve and prevent the Race Condition and its consequences, the thread module offers a Lock class which uses Semaphores to help threads synchronise. Semaphores are nothing but binary flags. 

The Lock class has two primary methods:

<li><b>acquire([blocking]):</b> The acquire method takes in the parameter blocking as either True or False. If a lock for a thread T1 was initiated with blocking as True, it will wait or remain blocked until the critical section of code is locked by another thread T2. Once the other thread T2 releases the lock, thread T1 acquires the lock and returns True.</li>
<li><b>release():</b> When the release method is called on the lock, it will unlock the lock and return True. Also, it will check if any threads are waiting for lock to be released. If there are, then it will allow exactly one of them to access the lock.</li>

In [ ]:
import threading
import time
import logging
import random

logging.basicConfig(level=logging.DEBUG, format='(%(threadName)-9s) %(message)s',)
                    
class Counter(object):
    def __init__(self, start = 0):
        self.lock = threading.Lock()
        self.value = start
    def increment(self):
        logging.debug('Waiting for a lock')
        self.lock.acquire()
        try:
            logging.debug('Acquired a lock')
            self.value = self.value + 1
        finally:
            logging.debug('Released a lock')
            self.lock.release()

def worker(c):
    for i in range(2):
        r = random.random()
        logging.debug('Sleeping %0.02f', r)
        time.sleep(r)
        c.increment()
    logging.debug('Done')

if __name__ == '__main__':
    counter = Counter()
    for i in range(2):
        t = threading.Thread(target=worker, args=(counter,))
        t.start()

    logging.debug('Waiting for worker threads')
    main_thread = threading.currentThread()
    for t in threading.enumerate():
        if t is not main_thread:
            t.join()
    logging.debug('Counter: %d', counter.value)

(Thread-5 (worker)) Sleeping 0.66
(Thread-6 (worker)) Sleeping 0.59
(MainThread) Waiting for worker threads
C:\Users\ADMIN\AppData\Local\Temp\ipykernel_5128\837421409.py:37: DeprecationWarning: currentThread() is deprecated, use current_thread() instead
  main_thread = threading.currentThread()
(Thread-6 (worker)) Waiting for a lock
(Thread-6 (worker)) Acquired a lock
(Thread-6 (worker)) Released a lock
(Thread-6 (worker)) Sleeping 0.59
(Thread-5 (worker)) Waiting for a lock
(Thread-5 (worker)) Acquired a lock
(Thread-5 (worker)) Released a lock
(Thread-5 (worker)) Sleeping 0.09
(Thread-5 (worker)) Waiting for a lock
(Thread-5 (worker)) Acquired a lock
(Thread-5 (worker)) Released a lock
(Thread-5 (worker)) Done
(Thread-6 (worker)) Waiting for a lock
(Thread-6 (worker)) Acquired a lock
(Thread-6 (worker)) Released a lock
(Thread-6 (worker)) Done


### Deadlocks

Another issue which arises when we deal with multiple locks is – Deadlocks. Deadlocks occur when locks are not released by threads due to various reasons.

In [ ]:
import threading

l = threading.Lock()
# Before the 1st acquire
l.acquire()
# Before the 2nd acquire
l.acquire()
# Now acquired the lock twice

In the above code, we call the acquire method twice but don’t release it after it is acquired for the first time. Hence, when Python sees the second acquire statement, it will go into the wait mode indefinitely as we never released the previous lock.

# Networking
Networking allows computer programs to communicate with each other, even if they are running on different machines. 
<br>For programs such as web browsers, networking is the essence of what they do, whereas for others networking adds additional dimensions to their functionality, for example, remote operation or logging, or the ability to retrieve or supply data to other machines. 
<br>Most networking programs work on  
<li>either a peer-to-peer basis (the same program runs on different machines), or 
<li>more commonly, a client/server basis (client programs send requests to a server)

https://www.youtube.com/watch?v=3QiPPX-KeSc

# The Socket Layer
In simple terms, sockets are a programmable interface to connections between programs, possibly running on different computers of a network. They allow data formatted as byte strings to be passed between processes and machines. 
<br>Sockets also form the basis and low-level “plumbing” of the Internet itself: all of the familiar higher-level Net protocols, like FTP, web pages, and email, ultimately occur over sockets. 
<br>Sockets are also sometimes called communications endpoints because they are the portals through which programs send and receive bytes during a conversation.

## Machine identifiers
Internet defines standard ways to name both a remote machine and a service provided by that machine. Within a script, the computer program to be contacted through a socket is identified by supplying a pair of values—the machine name and a specific port number on that machine

### 1. Machine names
A machine name may take the form of either a string of numbers separated by dots, called an IP address (e.g., **166.93.218.100**), or a more legible form known as a domain name (e.g., starship.python.net).

### 2. Port numbers
A port number is an agreed-upon numeric identifier for a given conversation. Because computers on the Net support a variety of services, port numbers are used to name a particular conversation on a given machine.


# The Protocol Layer
Although sockets form the backbone of the Internet, much of the activity that happens on the Net is programmed with protocols, which are higher-level message models that run on top of sockets.

The Internet Protocol standardize both message formats and socket port numbers:

<br>• Message formats provide structure for the bytes exchanged over sockets during conversations.
<br>• Port numbers are reserved numeric identifiers for the underlying sockets over which messages are exchanged.


### Port numbers reserved for common protocols

<table>
    <tr>
        <th>Protocol 
        <th>Common function                 
        <th>Port number 
        <th>Python module
    </tr>       
       
<tr><td>HTTP 
    <td>Web pages 
     <td>80 
         <td>http.client, http.server</tr>
<tr>
    <td>NNTP 
        <td>Usenet news 
            <td>119 
                <td>nntplib</tr>
<tr><td>FTP data default 
    <td>File transfers 
        <td>20 
            <td>ftplib</tr>
<tr><td>FTP control
    <td>File transfers 
        <td>21 
            <td>ftplib</tr>
<tr><td>SMTP 
    <td>Sending email 
        <td>25 
            <td>smtplib</tr>
<tr><td>POP3 
    <td>Fetching email 
        <td>110 
            <td>poplib</tr>
<tr><td>IMAP4 
    <td>Fetching email 
        <td>143 
            <td>imaplib</tr>
<tr><td>Finger 
    <td>Informational 
        <td>79 
            <td>n/a</tr>
<tr><td>SSH 
    <td>Command lines 
        <td>22 
            <td>n/a: third party</tr>
<tr><td>Telnet 
    <td>Command lines 
        <td>23 
            <td>telnetlib</tr>
    </table>

### Clients and servers
By defining standard port numbers for services, the Net naturally gives rise to a client/ server architecture. 
<li>On one side of a conversation, machines that support standard protocols perpetually run a set of programs that listen for connection requests on the reserved ports. 
<li>On the other end of a dialog, other machines contact those programs
to use the services they export.</li>

#### Server
A machine that hosts websites usually runs a web server program that constantly listens for incoming connection requests, on a socket bound to port 80. Often, the server itself does nothing but watch for requests on its port perpetually; handling
requests is delegated to spawned processes or threads.

#### Clients
Programs that wish to talk to this server specify the server machine’s name and port 80 to initiate a connection. For web servers, typical clients are web browsers like Firefox, Internet Explorer, or Chrome, but any script can open a client-side
connection on port 80 to fetch web pages from the server. The server’s machine name can also be simply “localhost” if it’s the same as the client’s.




Mark Lutz  - Programming Python-O'Reilly (2011)<br>
https://docs.oracle.com/cd/E19620-01/805-4041/6j3r8iu2g/index.html

### Server and Client code

In [ ]:
%%writefile echo-server.py
# \Sockets\echo-server.py
"""
Server side: open a TCP/IP socket on a port, listen for a message from
a client, and send an echo reply; this is a simple one-shot listen/reply
conversation per client, but it goes into an infinite loop to listen for
more clients as long as this server script runs; the client may run on
a remote machine, or on same computer if it uses 'localhost' for server
"""
from socket import * # get socket constructor and constants
myHost = '' # '' = all available interfaces on host
myPort = 50008 # listen on a non-reserved port number
sockobj = socket(AF_INET, SOCK_STREAM) # make a TCP socket object
sockobj.bind((myHost, myPort)) # bind it to server port number
sockobj.listen(5) # listen, allow 5 pending connects
while True: # listen until process killed
    connection, address = sockobj.accept() # wait for next client connect
    print('Server connected by', address) # connection is a new socket
    while True:
        data = connection.recv(1024) # read next line on client socket
        if not data: break # send a reply line to the client
        connection.send(b'Echo=>' + data) # until eof when socket closed
    
    connection.close()

Overwriting echo-server.py


In [ ]:
%%writefile echo-client.py
# \Sockets\echo-client.py
"""
Client side: use sockets to send data to the server, and print server's
reply to each message line; 'localhost' means that the server is running
on the same machine as the client, which lets us test client and server
on one machine; to test over the Internet, run a server on a remote
machine, and set serverHost or argv[1] to machine's domain name or IP addr;
Python sockets are a portable BSD socket interface, with object methods
for the standard socket calls available in the system's C library;
"""
import sys
from socket import * # portable socket interface plus constants
serverHost = 'localhost' # server name, or: 'starship.python.net'
serverPort = 50008 # non-reserved port used by the server

message = [b'Hello network world'] # default text to send to server
                                 # requires bytes: b'' or str,encode()
    
if len(sys.argv) > 1:
    serverHost = sys.argv[1] # server from cmd line arg 1
    if len(sys.argv) > 2: # text from cmd line args 2..n
        message = (x.encode() for x in sys.argv[2:])
        
sockobj = socket(AF_INET, SOCK_STREAM) # make a TCP/IP socket object
sockobj.connect((serverHost, serverPort)) # connect to server machine + port

for line in message:
    sockobj.send(line) # send line to server over socket
    data = sockobj.recv(1024) # receive line from server: up to 1k
    print('Client received:', data) # bytes are quoted, was `x`, repr(x)
    
sockobj.close() # close socket to send eof to server

Overwriting echo-client.py


In [ ]:
%%writefile thread-server.py
# \Sockets\thread-server.py
"""
Server side: open a socket on a port, listen for a message from a client,
and send an echo reply; echoes lines until eof when client closes socket;
spawns a thread to handle each client connection; threads share global
memory space with main thread; this is more portable than fork: threads
work on standard Windows systems, but process forks do not;
"""
import time, _thread as thread # or use threading.Thread().start()
from socket import * # get socket constructor and constants
myHost = '' # server machine, '' means local host
myPort = 50008 # listen on a non-reserved port number

sockobj = socket(AF_INET, SOCK_STREAM) # make a TCP socket object
sockobj.bind((myHost, myPort)) # bind it to server port number
sockobj.listen(5) # allow up to 5 pending connects

def now():
    return time.ctime(time.time()) # current time on the server

def handleClient(connection): # in spawned thread: reply
    time.sleep(5) # simulate a blocking activity
    while True: # read, write a client socket
        data = connection.recv(1024)
        if not data: break
        reply = 'Echo=>%s at %s' % (data, now())
        connection.send(reply.encode())
    connection.close()
        
def dispatcher(): # listen until process killed
    while True: # wait for next connection,
        connection, address = sockobj.accept() # pass to thread for service
        print('Server connected by', address, end=' ')
        print('at', now())
        thread.start_new_thread(handleClient, (connection,))
dispatcher()


Overwriting thread-server.py


# Python in HTML
Web Development is a vast field, and there are endless opportunities and things that we can do. With complexity and demand come requirements. When building dynamic web pages, we often have to perform tasks that require the assistance of some programming language such as Python or PHP.

Run Python script in HTML using Django
#### https://www.youtube.com/watch?v=LQTMqGns7Co
#### https://www.youtube.com/watch?v=zuxzE7--RYM

### Install Django
https://www.djangoproject.com/download/
<br>https://docs.djangoproject.com/en/4.0/intro/install/
<br>https://www.delftstack.com/howto/python/python-in-html/

In [ ]:
# Make a new directory
import os
os.mkdir("my_new_app1")
print("A new directory my_new_app1 created")


FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'my_new_app1'

In [ ]:
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011'

open cmd and chdir to ```my_new_app```

In [ ]:
os.chdir('./my_new_app1')

In [ ]:
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1'

set virtual environment using
```
python -m venv env
```

In [ ]:
!python -m venv env

activate virtual environment using
```
env\Scripts\activte
```

In [ ]:
!env\Scripts\activate

install Django using
```
pip install django
```

In [ ]:
!pip install django

  Using cached sqlparse-0.4.2-py3-none-any.whl (42 kB)


You should consider upgrading via the 'C:\Users\ADMIN\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Start new project using
```
django-admin startproject my_django_app
```

In [ ]:
!django-admin startproject my_django_app

change dirctory to my_django_app

In [ ]:
os.chdir('./my_django_app')
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app'

Start an app (you can create any number of app) using
```
python manage.py startapp hello
```

In [ ]:
!python manage.py startapp hello

Edit the file ```setting.py``` of my_django_app directory
<br> by adding
```
'hello'
```
<br> in the lsit ```INSTALLED_APPS``` as follows

```
INSTALLED_APPS=[
    'django.contrib.admin',
    .
    .
    .
    'django.cotrib.staticfiles',
    'hello',
    ]
```

In [ ]:
os.chdir("E:\VITBhopal\Jan 2022\CSE3011\my_new_app1\my_django_app")

In [ ]:
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app'

run server using cmd
```
python manage.py runserver
```

In [ ]:
!start python manage.py runserver  #start help in opening in new console

In [ ]:
# apply the migration using
# python manage.py migrate

In [ ]:
# create superuser by
# python manage.py createsuperuser
# user: lokesh
# email: ashokbecse2006@yahoo.com
# pwd: lokesh

In [ ]:
import os
os.chdir('./my_new_app/my_django_app')
os.getcwd()

FileNotFoundError: [WinError 3] The system cannot find the path specified: './my_new_app/my_django_app'

Change the directory to <b>hello directory
    <br>Edit the file ```views.py``` 

In [ ]:
import os
#os.chdir('..')
os.chdir('E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\hello')
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\hello'

# update views.py in hello directory
```
from django.shortcuts import render
# Create your views here.
def index(request):
    return render(request, 'hello/index.html')
```

In [ ]:
!notepad views.py

In [ ]:
import os
os.mkdir('templates')
os.chdir('./templates')
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\hello\\templates'

In [ ]:
import os
os.mkdir('hello')
os.chdir('./hello')
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\hello\\templates\\hello'

In [ ]:
f = open('index.html','w')
str = '''
<h1> Hello world</h1>
'''
f.write(str)
f.close()

In [ ]:
import os

os.chdir('E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\my_django_app')

os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011\\my_new_app1\\my_django_app\\my_django_app'

# update urls.py in my_django_app drive

following lines can be added<br>
from hello.views import index</br>
<br>path('hello/', index)</br>
```
"""my_django_app URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/4.0/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.contrib import admin
from django.urls import path
from hello.views import index

urlpatterns = [
    path('admin/', admin.site.urls),
    path('hello/', index)


]
```

In [ ]:
!notepad urls.py

Edit the url of browser as <br>
http://127.0.0.1:8000/hello/

# Download pages
Using Python to download files from the internet is super easy and possible using only standard library functions.

In [ ]:
import os
os.chdir("E:\VITBhopal\Jan 2022\CSE3011")
os.getcwd()

'E:\\VITBhopal\\Jan 2022\\CSE3011'

In [ ]:
# Python’s urllib library offers a range of functions designed to handle common URL-related tasks. 
# This includes parsing, requesting, and downloading files.
from urllib import request
# Define the remote file to retrieve
remote_url = 'https://www.google.com/robots.txt'
# Define the local filename to save data
local_file = 'local_copy.txt'
# Download remote and save locally
request.urlretrieve(remote_url, local_file)

('local_copy.txt', <http.client.HTTPMessage at 0x257a93eec20>)

In [ ]:
# The Python Requests module is a super-friendly library billed as “HTTP for humans.” 
# Offering very simplified APIs, Requests lives up to its motto for even high-throughput HTTP-related demands. 
# However, it doesn’t feature a one-liner for downloading files.
import requests
# Define the remote file to retrieve
remote_url = 'https://www.google.com/robots.txt'
# Define the local filename to save data
local_file = 'local_copy.txt'
# Make http request for remote file data
data = requests.get(remote_url)
# Save file data to local copy
with open(local_file, 'wb')as file:
    file.write(data.content)

In [ ]:
!pip install wget

  Using cached wget-3.2.zip (10 kB)
Using legacy 'setup.py install' for wget, since package 'wheel' is not installed.
    Running setup.py install for wget: started
    Running setup.py install for wget: finished with status 'done'


You should consider upgrading via the 'C:\Users\ADMIN\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [ ]:
# The wget Python library offers a method similar to urllib and attracts 
# a lot of attention due to its name being identical to the Linux wget command. 
import wget
# Define the remote file to retrieve
remote_url = 'https://www.google.com/robots.txt'
# Define the local filename to save data
local_file = 'local_copy.txt'
# Make http request for remote file data
wget.download(remote_url, local_file)

100% [................................................................................] 7240 / 7240

'local_copy (1).txt'

### About URLs
A web page is a file that is stored on another computer, a machine known as a web server. When you “go to” a web page, what is actually happening is that your computer, (the client) sends a request to the server (the host) out over the network, and the server replies by sending a copy of the page back to your machine.

The URL tells your browser where to find an online resource by specifying the server, directory and name of the file to be retrieved, as well as the kind of protocol that the server and your browser will agree to use while exchanging information (like HTTP, the Hypertext Transfer Protocol). 

The basic structure of a URL is
```
protocol://host:port/path?query
ex.
http://oldbaileyonline.org/static/Project.jsp
http://www.oldbaileyonline.org/browse.jsp?id=t17800628-33&div=t17800628-33

```

##### https://programminghistorian.org/en/lessons/working-with-web-pages

In [ ]:
# Opening URLs with Python
import urllib.request, urllib.error, urllib.parse

url = 'http://www.oldbaileyonline.org/browse.jsp?id=t17800628-33&div=t17800628-33'

response = urllib.request.urlopen(url)
webContent = response.read().decode('UTF-8')

print(webContent[0:300])















<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<title>Browse - Central Criminal Court</title>
	<meta http-equiv="content-type" content=


In [ ]:
# Saving a Local Copy of a Web Page
import urllib.request, urllib.error, urllib.parse

url = 'http://www.oldbaileyonline.org/browse.jsp?id=t17800628-33&div=t17800628-33'

response = urllib.request.urlopen(url)
webContent = response.read().decode('UTF-8')

f = open('obo-t17800628-33.html', 'w')
f.write(webContent)
f.close

<function TextIOWrapper.close()>

# CGI 


https://www.youtube.com/watch?v=Nrrz_0jEqGQ

https://www.youtube.com/watch?v=8LEWb-eF8TE



<li>start apache server in xmapp control panel
<li> go to browser and check 127.0.0.1:80 for apache server
<li>click on config tab on apache server and select httpd.config
<li> Search for option Indexes and ExecCGI is there if not add
<br>
    
    Options Indexes FollowSymLinks Includes ExecCGI
 <br>
<li> Search for AddHandler and add .py and .cgi
    <br>
        
            AddHandler cgi-script .cgi .pl .asp .py
<br>
<li> save the file and close it
<li> restart the apache from xampp control panel
<li> create a pycgi directory inside xampp/htdocs
<li> create .py file inside pycgi directory


In [ ]:
#!C:\Users\ADMIN\AppData\Local\Programs\Python\Python38\python.exe

print("Content-type:text/html\r\n\r\n")
print('<html>')
print('<head>')
print('<title>Hello World - First CGI Program</title>')
print('</head>')
print('<body>')
print('<h2>Hello World! This is my first CGI program</h2>')
print('</body>')
print('</html>')


save as hello.py
<br> open url in browser 127.0.0.1:80/pycgi/hello.py
<br> you will get
<br><b>Hello World! This is my first CGI program

In [ ]:
#!C:\Users\ADMIN\AppData\Local\Programs\Python\Python310\python.exe

# Import modules for CGI handling 
import cgi, cgitb 

# Create instance of FieldStorage 
form = cgi.FieldStorage() 

# Get data from fields
first_name = form.getvalue('first_name')
last_name  = form.getvalue('last_name')

print("Content-type:text/html\r\n\r\n")
print('<html>')
print('<head>')
print('<title>Hello World - Second CGI Program</title>')
print('</head>')
print('<body>')
print("<h2>Hello %s %s</h2>" % (first_name, last_name))
print('</body>')
print('</html>')


save as hello_get.py
<br> open url in browser 127.0.0.1:80/pycgi/hello_get.py?first_name=Ashok&last_name=Patel
<br> you will get
<br><b>Hello Ashok Patel

In [ ]:
<html>
<head>
<title>Get Response</title>
</head>

<body>
<form action = "/pycgi/hello_get.py" method = "get">
First Name: <input type = "text" name = "first_name">  <br />

Last Name: <input type = "text" name = "last_name" />
<input type = "submit" value = "Submit" />
</form>
<body>

create html page and passes two values using HTML FORM and submit button.

In [ ]:
<html>
<head>
<title>Get Response</title>
</head>

<body>
<form action = "/pycgi/hello_get.py" method = "post" target = "_blank" >
First Name: <input type = "text" name = "first_name">  <br />

Last Name: <input type = "text" name = "last_name" />
<input type = "submit" value = "Submit" />
</form>
<body>

# Other resources
https://www.youtube.com/watch?v=VMP1oQOxfM0

Mark Lutz - Programming Python-O'Reilly (2011)